plugins/plugin-java/che-plugin-java-ext-lang-client/src/main/java/org/eclipse/che/ide/ext/java/client/tree/library/JarFileNode.java

```diff
@@ -179,4 +198,55 @@ public boolean isContentGenerated() {
     public Path getProject() {
         return project;
     }
+
+    @Override
+    public void onFileOperation(FileEvent event) {
+        Path filePath = event.getFile().getLocation();
+        Path currentPath = getLocation();
+
+        if (!filePath.equals(currentPath)) {
+            return;
+        }
+
+        switch (event.getOperationType()) {
+            case OPEN: {
+                if (resourceChangeHandlerRegistration == null) {
+                    resourceChangeHandlerRegistration = eventBus.addHandler(ResourceChangedEvent.getType(), this);
+                }
+
+                break;
+            }
+
+            case CLOSE: {
+                if (resourceChangeHandlerRegistration != null) {
+                    resourceChangeHandlerRegistration.removeHandler();
+                }
+
+                break;
+            }
+        }
+    }
+
+    @Override
+    public void onResourceChanged(ResourceChangedEvent event) {
+        ResourceDelta delta = event.getDelta();
+        Path resourceLocation = delta.getResource().getLocation();
+
+        if (project.equals(resourceLocation) && REMOVED == delta.getKind()) {
```

Comment:
Check for delta kind at first, then for equality of resources. It will be faster than this condition.